consul-lambda-registrator/upsert_event.go

```diff
@@ -67,25 +74,74 @@ func (env Environment) storeServiceDefaults(e UpsertEvent) error {
 
 	// There is no need for CAS because we are completely regenerating the service
 	// defaults config entry.
-	_, _, err := env.ConsulClient.ConfigEntries().Set(serviceDefaults, e.writeOptions())
+	_, _, err := env.ConsulClient.ConfigEntries().Set(serviceDefaults, e.WriteOptions())
 
 	return err
 }
 
-func (e UpsertEvent) writeOptions() *api.WriteOptions {
-	var writeOptions *api.WriteOptions
-	if e.EnterpriseMeta != nil {
-		writeOptions = &api.WriteOptions{
-			Partition: e.EnterpriseMeta.Partition,
-			Namespace: e.EnterpriseMeta.Namespace,
+func (env Environment) upsertTLSData(e UpsertEvent) error {
+	if !env.IsManagingTLS() {
+		return nil
+	}
+
+	env.Logger.Debug("upserting mTLS data", "service", e.Name)
+
+	// Retrieve Consul root CA
+	// TODO: Should optimize this flow so that we only request the root CA once, not for every service.
+	// TODO: Retrieve the root CA IFF it is expiring "soon".
+	caRootList, _, err := env.ConsulClient.Agent().ConnectCARoots(nil)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve Consul root CA: %w", err)
+	}
+
+	// Use the first active root CA cert
+	var caRoot *api.CARoot
+	for _, root := range caRootList.Roots {
+		if root.ID == caRootList.ActiveRootID {
+			caRoot = root
+			break
 		}
 	}
+	if caRoot == nil {
+		return fmt.Errorf("failed to find an active CA root cert: %w", err)
+	}
+
+	// Retrieve the leaf for this service
+	leafCert, _, err := env.ConsulClient.Agent().ConnectCALeaf(e.Name, e.QueryOptions())
+	if err != nil {
+		return fmt.Errorf("failed to retrieve leaf cert for %s: %w", e.Name, err)
+	}
+
+	extData, err := json.Marshal(structs.ExtensionData{
+		PrivateKeyPEM: leafCert.PrivateKeyPEM,
+		CertPEM:       leafCert.CertPEM,
+		RootCertPEM:   caRoot.RootCertPEM,
+		TrustDomain:   caRootList.TrustDomain,
+		// TODO: cluster peering support
+	})
+	if err != nil {
+		return fmt.Errorf("failed to marshal extension data: %w", err)
+	}
+
+	// Create the info for this service
+	service := structs.Service{
+		Name:        e.Name,
+		Datacenter:  e.Datacenter,
+		TrustDomain: caRootList.TrustDomain,
+	}
+	if e.EnterpriseMeta != nil {
+		service.Namespace = e.EnterpriseMeta.Namespace
+		service.Partition = e.EnterpriseMeta.Partition
+	}
+	path := fmt.Sprintf("%s%s", env.ExtensionDataPath, service.ExtensionPath())
 
-	return writeOptions
+	// TODO: do we need to pass a context in here?.. like from the lambda entrypoint
+	// so that this call can be canceled if necessary.
```

Comment:
Probably. But this feels like a post beta thing. Maybe we should just make a ticket so we don't forget it?